src/executor.cc

```diff
@@ -110,30 +110,17 @@ namespace detail {
 				}
 			}
 
-			MPI_Status status;
-			int flag;
-			MPI_Message msg;
-			MPI_Improbe(MPI_ANY_SOURCE, mpi_support::TAG_CMD, MPI_COMM_WORLD, &flag, &msg, &status);
-			if(flag == 1) {
-				int frame_bytes;
-				MPI_Get_count(&status, MPI_BYTE, &frame_bytes);
-				unique_frame_ptr<command_frame> frame(from_size_bytes, static_cast<size_t>(frame_bytes));
-				MPI_Mrecv(frame.get_pointer(), frame_bytes, MPI_BYTE, &msg, &status);
-				command_queue.push(std::move(frame));
-
-				if(!m_first_command_received) {
-					m_metrics.initial_idle.pause();
-					m_metrics.device_idle.resume();
-					m_first_command_received = true;
-				}
-			}
-
-			if(m_jobs.size() < MAX_CONCURRENT_JOBS && !command_queue.empty()) {
-				if(!handle_command(*command_queue.front())) {
-					// In case the command couldn't be handled, don't pop it from the queue.
-					continue;
+			if(m_jobs.size() < MAX_CONCURRENT_JOBS) {
+				// TODO: Double-buffer command queue?
+				// FIXME: Don't hold lock while calling handle_command (it's annoying b/c handling can fail)
+				std::unique_lock lk(m_command_queue_mutex);
+				if(!m_command_queue.empty()) {
+					if(!handle_command(m_command_queue.front())) {
+						// In case the command couldn't be handled, don't pop it from the queue.
+						continue;
+					}
+					m_command_queue.pop();
```

Comment:
I have no intuition whether unlocking around `handle_command` would benefit us, but it could be implemented like this:
```suggestion
				std::unique_lock lk(m_command_queue_mutex);
				if(!m_command_queue.empty()) {
					auto pkg = std::move(m_command_queue.front());
					lk.unlock();
					const auto handled = handle_command(pkg);
					lk.lock();
					if (handled) {
						m_command_queue.pop();
					} else {
						// In case the command couldn't be handled, don't pop it from the queue.
						m_command_queue.front() = std::move(pkg);
						continue;
					}
```

Reply:
Thanks, I've added it. It's probably not going to do much, but I still think it's cleaner that way.